CIP-0118/README-implicit.md

```diff
@@ -0,0 +1,895 @@
+
+Omitting `TxStructure` and mapping per era should also be acceptable, if desired:
+  ```
+  fromTxZones :: TxZones era -> StrictSeq (StrictSeq (Tx era))
+  toTxZones :: StrictSeq (StrictSeq (Tx era)) -> TxZones era
+  ```
+
+An example of how this looks, in full, for two eras supporting different structures:
+  ```
+  instance Crypto c => EraSegWits (ConwayEra c) where
+    type TxStructure (ConwayEra c) = StrictSeq
+    type TxZones (ConwayEra c) = AlonzoTxSeq (ConwayEra c)
+    fromTxZones = txSeqTxns
+    toTxZones = AlonzoTxSeq
+
+  instance Crypto c => Core.EraSegWits (BabelEra c) where
+    type TxStructure (BabelEra c) = Compose StrictSeq StrictSeq
+    type TxZones (BabelEra c) = BabelTxZones (BabelEra c)
+    fromTxZones = Compose . txZonesTxns
+    toTxZones = BabelTxZones . getCompose
+  ```
+
+For more information, search for `CIP-0118#block-structure-0` in the codebase, which can be found [here](https://github.com/willjgould/cardano-ledger/tree/wg/babel-fees-prototyping-babel-erafirstrule).
+Please note that the most up-to-date prototype branch is `babel-fees-prototyping-babel-erafirstrule`.
+
+#### Transaction structure
+
+Probably no changes
+
+#### Ledger State
+
+Probably no changes
+
+#### Ledger State
+
+The changes follow the specification changes
+
+#### CBOR
+
+Probably no changes
+
+### New Plutus Version
+
+Probably don't need one 
+
+### Consensus
+
+We do not expect that any functional changes are required for consensus - the only
+change needed is likely the updating of any references to the block body type,
+which will be `List (List Tx)` instead of `List Tx`.
+
+### Mempool
+
+The mempool must switch to operating on zones instead of singleton transactions.
+It must also support adding _part of a zone_ to a block in the case of phase-2
+validation failure. That is, if a phase-2 invalid transaction is encountered
+during zone validation, subsequent transactions are not validated, and only the transactions
+validated up to and including this first phase-2 failing one are included in the
+zone. This partial zone is then included in the block.
+
+### Network Requirements and Changes
+
+Zones replace singleton transactions as units sent across the network. Zones must
+therefore have IDs (e.g. zone hashes) for a node to determine whether it has
+already downloaded the corresponding zone based on an ID it is observing on the network.
+To maintain network functionality, zone sizes must be limited by the current `maxTxSize`
+protocol parameter.
+
+**Decision Required**. Is this an acceptable constraint for the usecases we
+want to support?
+
+### CLI and API
+
+The node CLI should support the construction of the new type of transaction body,
+as well as the construction and (local) validation of a zone.
+The transaction balancing must also be modified in a way that lets the user explicitly 
+(locally) specify the amounts they want to be missing or extra in a transaction. 
```

Comment:
> how transaction zone creation will be happening offline and the interdependency of transaction within the zone (i.e. the DAG relationship).

The transactions are constructed as usual. The only new field a user must specify is requiredTxs, which is needed to ensure inclusion of other transactions in the same zone, such as for the purpose of allowing Plutus scripts to inspect them. Transactions do not depend on each other inside a zone in any way *unless* one specifies another's ID in the `requiredTxs` field. This is similar to how the top level transaction in Swaps depends on its swaps, however, this relationship may be in the form of any DAG. 

> I am really not liking the change to block serialization and requirement of consensus to specially treat those zones in the mempool, but maybe I am overthinking it and in the end it won't be too mush of a problem once we get to actual implementation. There will have to be a new mechanism invented for communication between consensus and ledger for consensus to know which transactions need to be dropped

@willjgould  might have some better idea of this

> inability for plutus scripts to see other transactions where the missing value is coming from. Maybe it doesn't matter at all and script writer won't care about this, so that's just my personal opinion

I addressed this in the updated design - transaction `tx` can specify `requiredTxs : TxId |-> Bool` and use the boolean to toggle whether they want only a phase-1 check that the transaction with a given ID is present in the zone, or they want Plutus scripts in `tx` to see the full required transactions. This lets scripts see into other transaction bodies, but does not force this. For example, a balanced transaction with no required Txs can run old-version scripts inside a zone, but a transaction that requires it will have its (new-version) Plutus scripts inspect its contents. This is also cheaper in terms of ExUnits

> Can't that information be computed automatically from the transaction? Calculating consumed - produced on a single transaction in the zone will give the amount that need to be balanced out, no?

Yes, this is just for ease of handling transaction construction from a user's perspective

Reply:
> I am not sure if this is better or worse than just having lists of lists, but so far we didnt foresee a problem with the change

You didn't see this because that is the work that would be coming from the consensus side. Support from ledger will be needed, but that is why you did not run into this problem yet.

> This lets scripts see into other transaction bodies, but does not force this.

I am not really sure that this is a good idea. This approach of sometimes allowing the script to see only some of the transactions in the zone feels a bit arbitrary. Let's say you have a script that requires you to specify other transactions in the zone, but there is only one occasion when you can supply all of them, because there will be other transactions that depends on the transaction you are building in the zone. For example, only the final transaction in the zone has this ability to supply all of the transactions in the zone, therefore such script could only be used in the final transaction. From the perspective of balancing transactions there are no real dependencies between transactions in the zone, that process is 100% commutative. Therefore I am not convinced that `requiredTxs` is a good idea at all. 

> Yes, this is just for ease of handling transaction construction from a user's perspective

Yep, makes sense.


> Transactions do not depend on each other inside a zone in any way unless one specifies another's ID in the requiredTxs field

There is implicit DAG dependency through the collateral requirement.

Reply:
> There is implicit DAG dependency through the collateral requirement.

No, this is no longer true. Only cover your own collateral. This was a legacy requirement from previous design, as I realized. Because that design allowed for transactions to spend each-others outputs within a single zone.

> For example, only the final transaction in the zone has this ability to supply all of the transactions in the zone, therefore such script could only be used in the final transaction. 

Yea, this is the DAG dependency. If you want to show a script a transaction, that transaction has to have been constructed already. This is similar to the top-level transaction in your case showing swaps to the scripts in it, but not the other way around. However, scripts aren't forced to see those other transactions.

Reply:
> You didn't see this because that is the work that would be coming from the consensus side. Support from ledger will be needed, but that is why you did not run into this problem yet.

Yeah, perhaps the `zoneID` solution would be better.

Reply:
> Yeah, perhaps the zoneID solution would be better.

ZoneId won't solve this, as far as I can imagine this solution. Ledger will still have to do the validation of each individual transaction in the zone and then the zone as a whole. mempool will need to be able to get the information of validity individual transaction in the zone in order to make other decisions about the zone, like at which piint to drop transactions. Moreover, ZoneId removes the guarantee that mempool will receive all of the transactions in the zone, unless requirement is to submit all necessary transactions with the same ZoneId at the same time. The the question is what happens with the ZoneId when transactions from the zone need to be dropped, that would be a problem, unless there is no connection between the ZoneId and the contents of the zone (i.e. ZoneId is just some random token) At that point the question is who is responsible for generating the ZoneId?

In any case, the ZoneId design is just not clear to me.

Reply:
> I think it should be possible to replace this with adding zoneID to the Tx structure. Instead of placing lists of transactions in blocks, a node would add such a field (which would be computed from the list coming across the network). The ZONE rule would still be needed, though. I am not sure if this is better or worse than just having lists of lists, but so far we didnt foresee a problem with the change

We could add a ZoneId, require that all txs in a zone are sent sequentially, and some sort of delimiting information (`eof`-like signal, number of expected txs in zone, or something). Certainly, converting to some mechanism without lists of lists of transactions shouldn't pose a problem. But at the same time, as I think @lehins is saying, any calculations regarding the dropping of transactions still need to be done, so I'm not sure this approach would be fruitful towards addressing the concern here.

So I suppose our options are to either extend the prototype to squash the concern directly, or just absorb the uncertainty into the cost of the implementation.